src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -4471,7 +4479,7 @@ class RGWBucketSyncSingleEntryCR : public RGWCoroutine {
 	      pretty_print(sc->env, "Syncing object s3://{}/{} in sync from zone {}\n",
 			   bs.bucket.name, key, zone_name);
 	    }
-            call(data_sync_module->sync_object(dpp, sc, sync_pipe, key, versioned_epoch,
+            call(data_sync_module->sync_object(dpp, sc, sync_pipe, key, timestamp, versioned_epoch,
```

Comment:
i think this timestamp comes from bilog, but we want to fetch the mtime from the source object

consider a series of object overwrites on the source zone that generate bilog entries at time t1, t2, t3. when the destination goes to sync that first bilog entry for t1, it fetches the latest upload

so we'll need to wait until fetch_remote_obj() has read the object's mtime before we can generate the version id. but looking at https://github.com/ceph/ceph/blob/c4baaa2/src/rgw/driver/rados/rgw_rados.cc#L4261, we don't parse the mtime until `complete_request()`. that might be too late, because we've already written the target object data at that point

Reply:
@cbodley right, we don't have access to the source's mtime until much later. so, I thought using bilog timestamp might not be a bad idea, but I see your point.

Reply:
I remember discussing the challenge of maintaining consistency on archive zone involving concurrent overwrites in this thread long ago:
https://github.com/ceph/ceph/pull/30667#discussion_r331130424

Reply:
i'd be worried about making any changes in fetch_remote_obj(). instead, we could generate the version after reading src object's mtime in RGWObjFetchCR() within data sync. we already call RGWStatRemoteObjCR() which gives the object mtime there.

Reply:
but the call to `conn->get_obj()` in `fetch_remote_obj` will still replace the object being synced, with the latest object if it was overwritten. this was a design limitation that was discussed long back in https://github.com/ceph/ceph/pull/30667#discussion_r331130424

for the archive case alone, I could try and add a separate `fetch_remote_obj` method to fetch the object first, generate the version and then write the `dest`, if that is possible. @cbodley what do you think?

Reply:
> we already call RGWStatRemoteObjCR() which gives the object mtime there

that (hopefully) doesn't happen for every object - we should only call stat in special bucket sync policy configurations. and that itself is racy, because the object could be overwritten between the HEAD and GET requests

> i'd be worried about making any changes in fetch_remote_obj()

i also don't want extra complexity there just for archive zone, but that seemed inevitable with this design

i was hoping we could find a way for `conn->get_obj()` to make that mtime available to the `cb` instead of `conn->complete_request()`, so that RGWRadosPutObj could calculate the version id before [processor.prepare()](https://github.com/ceph/ceph/blob/c4baaa2/src/rgw/driver/rados/rgw_rados.cc#L4214])

but ultimately we'd need that version id before we call [get_obj_state()](https://github.com/ceph/ceph/blob/c4baaa2/src/rgw/driver/rados/rgw_rados.cc#L4235C11-L4235C24) to read our local timestamp. we need to send that timestamp in an `If-Modified-Since` request header, so we can't depend on reading the `rgwx-mtime` response header first. this seems like a chicken/egg situation